src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala

```diff
@@ -0,0 +1,346 @@
+
+  private def sampleAvro(sc: ScioContext,
+                         input: String,
+                         output: String,
+                         sampleClass: Class[_],
+                         samplePct: Float,
+                         seed: Option[Int]): Future[Tap[GenericRecord]] = {
+    val schema = AvroIO.getAvroSchemaFromFile(input.toString)
+    val outputParts =
+      if (output.endsWith("/")) output + "part*" else output + "/part*"
+    if (FileStorage(outputParts).isDone) {
+      // sampling done previously - just reuse
+      log.info(s"Reuse previous sample at $outputParts")
+      Taps().avroFile(outputParts, schema = schema)
+    } else {
+      log.info(s"Will sample from: $input, output will be $output")
+      @transient lazy val sampleSchema = sampleClass.
+        asInstanceOf[Class[GenericRecord]].newInstance().getSchema
+      val (s, n) = samplePctToInt(samplePct, 100)
+      val r = sc.avroFile[GenericRecord](input, schema)
+        .flatMap { e =>
+          val hash =
+            hashAvroField(e, sampleSchema.getFields.asScala, kissHashFun(seed).newHasher()).hash()
+          diceElement(e, hash, s.toInt, n)
+        }
+        .saveAsAvroFile(output, schema = schema)
+      sc.close()
+      r
+    }
+  }
+
+  @tailrec
+  private def samplePctToInt(samplePct: Float, n: Int): (Float, Int) = {
+    if (samplePct < 1.0) {
+      samplePctToInt(samplePct * 10, n * 10)
+    } else {
+      (samplePct, n)
+    }
+  }
+
+  def singleInput(argv: Array[String]): Future[Tap[_]] = {
+    val (sc, args) = ContextAndArgs(argv)
+    val samplePct = args("sample").toFloat
+    require(samplePct > 0.0F && samplePct < 100.0F,
+      "Sample percentage should be between (0.0, 100.0)")
+
+    val input = args("input")
+    val output = args("output")
+    val sampleAvroSchemaClass = args.optional("avro-schema")
+    val fields = args.list("fields")
+    val seed = args.optional("seed")
+
+    if (parseAsBigQueryTable(input).isDefined) {
+      require(parseAsBigQueryTable(output).isDefined,
+        s"Input is a BigQuery table `${input}`, output should be a BigQuery table too," +
+          s"but instead it's `${output}`.")
+      require(fields.nonEmpty && sampleAvroSchemaClass.isEmpty,
+        "Input is a BigQuery table, please specify fields to sample on, via the --fields option.")
+      val inputTbl = parseAsBigQueryTable(input).get
+      val outputTbl = parseAsBigQueryTable(output).get
+      sampleBigQueryTable(sc, inputTbl, outputTbl, fields, samplePct, seed.map(_.toInt))
+    } else if (parseAsURI(input).isDefined) {
+      // right now only support for avro
+      require(parseAsURI(output).isDefined,
+        s"Input is a URI: `${input}`, output should be a URI too, but instead it's `${output}`.")
+      require(fields.isEmpty && sampleAvroSchemaClass.isDefined,
+        "Input is a URI, please specify sample schema via the --avro-schema option.")
+      val sampleClass = Class.forName(sampleAvroSchemaClass.get)
```

Comment:
Just double check if I fully understand this. For Avro you don't specify `fields: List[String]`, but instead a trimmed schema containing only fields you want to sample? This will make it hard to use from command line right?
How about stealing the schema projection logic from parquet-avro?